Mathlib/NumberTheory/ModularForms/EisensteinSeries/E2/Defs.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2025 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+
+module
+
+public import Mathlib.Analysis.CStarAlgebra.Classes
+public import Mathlib.NumberTheory.LSeries.RiemannZeta
+public import Mathlib.NumberTheory.ModularForms.EisensteinSeries.Summable
+public import Mathlib.NumberTheory.ModularForms.EisensteinSeries.Defs
+public import Mathlib.Topology.Algebra.InfiniteSum.ConditionalInt
+
+/-!
+# Eisenstein Series E2
+
+We define the Eisenstein series `E2` of weight `2` and level `1` as a limit of partial sums
+over non-symmetric intervals.
+
+-/
+
+open UpperHalfPlane hiding I
+
+open ModularForm EisensteinSeries Matrix.SpecialLinearGroup Filter Complex MatrixGroups
+  SummationFilter Real
+
+@[expose] public noncomputable section
+
+namespace EisensteinSeries
+
+/-- This is an auxilary summand used to define the Eisenstein serires `G2`. -/
+def e2Summand (m : ℤ) (z : ℍ) : ℂ := ∑' n, eisSummand 2 ![m, n] z
+
+lemma e2Summand_summable (m : ℤ) (z : ℍ) : Summable (fun n ↦ eisSummand 2 ![m, n] z) := by
+  apply (linear_right_summable z m (k := 2) (by grind)).congr
+  simp [eisSummand]
+
+@[simp]
+lemma e2Summand_zero_eq_two_riemannZeta_two (z : ℍ) : e2Summand 0 z = 2 * riemannZeta 2 := by
+  simpa [e2Summand, eisSummand] using
+    (two_mul_riemannZeta_eq_tsum_int_inv_pow_of_even (k := 2) (by grind) (by simp)).symm
+
+lemma e2Summand_even (z : ℍ) : (e2Summand · z).Even := by
+  intro n
+  simp only [e2Summand, ← tsum_comp_neg (fun a ↦ eisSummand 2 ![-n, a] z)]
+  apply tsum_congr (fun b ↦ ?_)
+  simp only [eisSummand, Fin.isValue, Matrix.cons_val_zero, Matrix.cons_val_one,
+    Matrix.cons_val_fin_one, Int.reduceNeg, zpow_neg, Int.cast_neg, neg_mul, inv_inj]
+  rw_mod_cast [Int.cast_neg]
+  ring
+
+/-- The Eisenstein series of weight `2` and level `1` defined as the conditional sum
+of `m` in `[N,N]` of `e2Summand m`. This sum over symmetric intervals is handy in showing it is
+Summable. -/
+def G2 : ℍ → ℂ := fun z ↦ ∑'[symmetricIcc ℤ] m, e2Summand m z
+
+/-- The normalised Eisenstein series of weight `2` and level `1`.
+Defined as `(1 / 2 * ζ(2)) * G2 `. -/
+def E2 : ℍ → ℂ := (1 / (2 * riemannZeta 2)) • G2
+
+/-- This function measures the defect in `G2` being a modular form. -/
+def D2 (γ : SL(2, ℤ)) : ℍ → ℂ := fun z ↦ (2 * π * I * γ 1 0) / (denom γ z)
+
+lemma D2_one : D2 1 = 0 := by
+  ext z
+  simp [D2]
+
+private lemma denom_aux (A B : SL(2, ℤ)) (z : ℍ) : ((A * B) 1 0) * (denom B z) =
+    (A 1 0) * B.1.det + (B 1 0) * denom (A * B) z := by
+  have h0 := Matrix.two_mul_expl A.1 B.1
+  simp only [Fin.isValue, coe_mul, h0.2.2.1, Int.cast_add, Int.cast_mul, ModularGroup.denom_apply,
+    Matrix.det_fin_two B.1, Int.cast_sub, ← map_mul, h0.2.2.2]
+  ring
+
```

Comment:
The following is entirely cosmetic, but it doesn't half make the infoview more readable:
```suggestion

local notation "φ" => Matrix.SpecialLinearGroup.map (Int.castRingHom ℝ) in
```